src/Core/OrganizationFeatures/OrganizationSponsorships/FamiliesForEnterprise/Cloud/CloudSyncSponsorshipsCommand.cs

```diff
@@ -0,0 +1,122 @@
+    {
+        private readonly IOrganizationRepository _organizationRepository;
+        private readonly IOrganizationSponsorshipRepository _organizationSponsorshipRepository;
+        private readonly IOrganizationUserRepository _organizationUserRepository;
+        private readonly ISendSponsorshipOfferCommand _sendSponsorshipOfferCommand;
+
+
+        public CloudSyncSponsorshipsCommand(
+        IOrganizationRepository organizationRepository,
+        IOrganizationSponsorshipRepository organizationSponsorshipRepository,
+        IOrganizationUserRepository organizationUserRepository,
+        ISendSponsorshipOfferCommand sendSponsorshipOfferCommand)
+        {
+            _organizationRepository = organizationRepository;
+            _organizationUserRepository = organizationUserRepository;
+            _organizationSponsorshipRepository = organizationSponsorshipRepository;
+            _sendSponsorshipOfferCommand = sendSponsorshipOfferCommand;
+        }
+
+        public async Task<OrganizationSponsorshipSyncData> SyncOrganization(Organization sponsoringOrg, IEnumerable<OrganizationSponsorshipData> sponsorshipsData)
+        {
+            if (sponsoringOrg == null)
+            {
+                throw new BadRequestException("Failed to sync sponsorship - missing organization.");
+            }
+            if (!sponsorshipsData.Any())
+            {
+                throw new BadRequestException("Failed to sync sponsorship - missing sponsorships.");
+            }
+
+            var existingSponsorships = await _organizationSponsorshipRepository.GetManyBySponsoringOrganizationAsync(sponsoringOrg.Id);
+            var existingSponsorshipsDict = existingSponsorships.ToDictionary(i => i.SponsoringOrganizationUserId);
+
+            var sponsorshipsToUpsert = new List<OrganizationSponsorship>();
+            var sponsorshipIdsToDelete = new List<Guid>();
+
+            foreach (var selfHostedSponsorship in sponsorshipsData)
+            {
+                var requiredSponsoringProductType = StaticStore.GetSponsoredPlan(selfHostedSponsorship.PlanSponsorshipType)?.SponsoringProductType;
+                if (requiredSponsoringProductType == null ||
+                    StaticStore.GetPlan(sponsoringOrg.PlanType).Product != requiredSponsoringProductType.Value)
+                {
+                    throw new BadRequestException("Specified Organization does not support this type of sponsorship.");
+                }
+
+                var cloudSponsorship = existingSponsorshipsDict[selfHostedSponsorship.SponsoringOrganizationUserId];
+                if (cloudSponsorship == null)
+                {
+                    if (selfHostedSponsorship.ToDelete && selfHostedSponsorship.LastSyncDate == null)
+                    {
+                        continue; // prevent invalid sponsorships in cloud. These should have been deleted by self hosted
+                    }
+                    if (!sponsoringOrg.Enabled)
```

Comment:
I'm not sure if we need one here. The grace period on the webhook was to prevent us from deleting sponsorships but this check only stops us from adding new ones. If the org does restart they will continue to sync and we can add those then

Reply:
Self hosted licenses stop working 2 months after the org is disabled. We should at least match that to allow full functionality during the grace period. I suggest three just to make it consistent across the feature.